libpromises/evalfunction.c

```diff
@@ -8450,6 +8451,84 @@ static FnCallResult FnCallNetworkConnections(EvalContext *ctx, ARG_UNUSED const
 
 /*********************************************************************/
 
+static FnCallResult FnCallFileReadable(ARG_UNUSED EvalContext *const ctx,
+                                       ARG_UNUSED const Policy *const policy,
+                                       const FnCall *const fp,
+                                       const Rlist *finalargs)
+{
+    assert(fp != NULL);
+    assert(fp->name != NULL);
+
+    if (finalargs == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Function %s requires path as first argument",
+            fp->name);
+        return FnFailure();
+    }
+    const char *const path = RlistScalarValue(finalargs);
+
+    finalargs = finalargs->next;
+    const long timeout = (finalargs == NULL) ? 0L
+                       : IntFromString(RlistScalarValue(finalargs));
+    assert(timeout >= 0);
+
+    FILE *const file = fopen(path, "r");
```

Comment:
Generally I don't think using the file stream (`FILE`) API is a good choice. It's not needed and it hides certain details. Most importantly, AFAICT, even this or even `open()` can potentially hang forever. So applying the timeout only to the `select()` below is too late.

Reply:
Also, this did not work on Windows. And Windows does not have the `poll` function where I can listen for events. 
```
   debug: File C:\test1.txt is not readable: Could not poll event: An operation was attempted on something that is not a socket.
```
Any ideas on how to do this?

@vpodzime do you think it would be better to not block at all?

Reply:
@larsewi @vpodzime I guess using `open` with `O_RDONLY` and `O_NONBLOCK` is what we want. Then you can only worry about timeout in `select()` - `open()`, `read()` and `close()` will be non-blocking.

Reply:
Well, my `man:open(2)` says:
>              Note that this flag has no effect for regular files and block devices; that is, I/O operations will (briefly) block when device activity is required, regardless of whether O_NONBLOCK is  set.
>              Since O_NONBLOCK semantics might eventually be implemented, applications should not depend upon blocking behavior when specifying this flag for regular files and block devices.

Not sure how it works with regular files on NFS, for example.

Reply:
> And Windows does not have the poll function where I can listen for events.

AFAICT, Windows doesn't have `poll()` and `select()` only works on sockets. We have some workarounds for that in the code already.

An option is to create a thread and try to `join()` it with a timeout and detach it if it fails.